src/runtime/stream.rs

```diff
@@ -0,0 +1,293 @@
+        let stream = Self::connect(address).await?;
+        Ok(stream)
+    }
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self> {
+        let stream = Self::connect_timeout(address, timeout).await?;
+        Ok(stream)
+    }
+}
+
+async fn try_connect<C: AsyncConnect>(address: &SocketAddr, timeout: Duration) -> Result<C> {
+    // The URI options spec requires that the default connect timeout is 10 seconds, but that 0
+    // should indicate no timeout.
+    let stream = if timeout == Duration::from_secs(0) {
+        C::connect(address).await?
+    } else {
+        C::connect_timeout(address, timeout).await?
+    };
+
+    Ok(stream)
+}
+
+async fn connect_stream<C: AsyncConnect>(
+    address: &StreamAddress,
+    connect_timeout: Option<Duration>,
+) -> Result<C> {
+    let timeout = connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT);
+
+    let mut socket_addrs: Vec<_> = address.to_socket_addrs()?.collect();
+
+    if socket_addrs.is_empty() {
+        return Err(ErrorKind::NoDnsResults(address.clone()).into());
+    }
+
+    // After considering various approaches, we decided to do what other drivers do, namely try each
+    // of the addresses in sequence with a preference for IPv4.
+    socket_addrs.sort_by_key(|addr| if addr.is_ipv4() { 0 } else { 1 });
+
+    let mut connect_error = None;
+
+    for address in &socket_addrs {
+        connect_error = match try_connect(address, timeout).await {
+            Ok(stream) => return Ok(stream),
+            Err(err) => Some(err),
+        };
+    }
+
+    Err(connect_error.unwrap_or_else(|| ErrorKind::NoDnsResults(address.clone()).into()))
+}
+
+impl AsyncStream {
+    /// Creates a new Tokio TCP stream connected to the server as specified by `options`.
+    #[cfg(feature = "tokio-runtime")]
+    pub(super) async fn connect_tokio(options: StreamOptions) -> Result<Self> {
+        use std::sync::Arc;
+
+        use tokio::net::TcpStream;
+        use tokio_rustls::TlsConnector;
+        use webpki::DNSNameRef;
+
+        let inner: TcpStream = connect_stream(&options.address, options.connect_timeout).await?;
+        inner.set_nodelay(true)?;
+
+        let inner = AsyncStreamInner::Tokio(inner);
+
+        // If there are TLS options, wrap the inner stream with rustls.
+        match options.tls_options {
+            Some(cfg) => {
+                let name = DNSNameRef::try_from_ascii_str(&options.address.hostname)?;
+                let mut tls_config = cfg.into_rustls_config()?;
+                tls_config.enable_sni = true;
+
+                let connector: TlsConnector = Arc::new(tls_config).into();
+                let session = connector.connect(name, inner).await?;
+
+                Ok(Self::Tls(session))
+            }
+            None => Ok(Self::Tcp(inner)),
+        }
+    }
+
+    #[cfg(feature = "async-std-runtime")]
+    pub(super) async fn connect_async_std(options: StreamOptions) -> Result<Self> {
```

Comment:
This is a copy of the sync `Stream::connect` with the only changes being using async-std's `TcpStream` and tokio_rustls's `TlsStream` instead of those from `std` and `rustls` respectively and making the function async and awaiting where necessary.